Support proper parsing of IPs with leading zeros

arkadeepsen opened the pull request:
Changed all ParseIP and ParseCIDR functions of net package to ParseIPSloppy and ParseCIDRSloppy of k8s.io/utils/net package respectively for k8s API objects. This is done to allow IPs and CIDRs with leading zeros to be parsed without any error.

Additionally have added conversion of any IP in string format to net.IP format and then conversion to string format again to remove any leading zeros.

Following set of IP strings are validated:
- `service.Spec.ClusterIP`
- `service.Spec.ClusterIPs`
- `service.Spec.ExternalIPs`
- `service.Status.LoadBalancer.Ingress[*].IP`
- `node.Status.Addresses[*].Address` (for address types `NodeInternalIP` and `NodeExternalIP`)
- `endpointSlice.Endpoints[*].Addresses`
- `pod.Status.PodIP`
- `pod.Status.PodIPs[*].IP`

Signed-off-by: Arkadeep Sen <arsen@redhat.com>

Fixes: https://issues.redhat.com/browse/OCPBUGS-463

arkadeepsen commented:
/assign @martinkennelly

arkadeepsen commented:
/retest-failed

npinaeva commented:
Hey @arkadeepsen, as we discussed, we need to add ips validation for all k8s objects that can set ips (and CIDRs)
So I'd suggest to have a list of these objects, and then create parsing functions, to make sure we parse every ip/cidr before working with it.
So we would have functions like `util.GetClusterIPs(service)` and it should return properly parsed ips, then we won't need duplicate code like this https://github.com/ovn-org/ovn-kubernetes/pull/3138/files#diff-bb96390aa4d292c5ff6e2bc554446182b8abd55e12ba2a0290853b92aa4e1155R417-R426
wdyt?

arkadeepsen commented:
> Hey @arkadeepsen, as we discussed, we need to add ips validation for all k8s objects that can set ips (and CIDRs) So I'd suggest to have a list of these objects, and then create parsing functions, to make sure we parse every ip/cidr before working with it. So we would have functions like `util.GetClusterIPs(service)` and it should return properly parsed ips, then we won't need duplicate code like this https://github.com/ovn-org/ovn-kubernetes/pull/3138/files#diff-bb96390aa4d292c5ff6e2bc554446182b8abd55e12ba2a0290853b92aa4e1155R417-R426 wdyt?

Makes more sense @npinaeva. I'll make the appropriate changes.

arkadeepsen commented:
/retest-failed

arkadeepsen commented:
/retest-failed

arkadeepsen commented:
/retest-failed

npinaeva commented on go-controller/pkg/cni/cniserver_test.go:
```diff
@@ -17,11 +16,14 @@ import (
 	"testing"
 
+	"k8s.io/client-go/kubernetes"
+
```
remove newline?

arkadeepsen replied:
Sure. Will remove it.

npinaeva commented:
I am a bit confused by changes like 
https://github.com/ovn-org/ovn-kubernetes/pull/3138/commits/c48d85433ed662d261679a12febea741812d0d82#diff-bb96390aa4d292c5ff6e2bc554446182b8abd55e12ba2a0290853b92aa4e1155L429
and then
https://github.com/ovn-org/ovn-kubernetes/pull/3138/commits/92b8cd3f7add1cb845cebfdf923303e98dc91586#diff-bb96390aa4d292c5ff6e2bc554446182b8abd55e12ba2a0290853b92aa4e1155R429
you basically did the change in the first commit and reverted it in the next. Can you please rebase commits so that they have final changes?
Also the commits as they are right now are bit difficult to review - maybe having commit that replaces net.Parse functions with parseSloppy and nothing else, and then next commits for changes related to objects you listed in the description will be easier to review?
Also, can you add some comments for unit test changes you did? I don't see why this https://github.com/ovn-org/ovn-kubernetes/pull/3138/commits/4e83aaedf2b49815ca43bc05388ad2df3252ec54 change is required - it looks for me like ClusterIP is the right service type

arkadeepsen commented:
> I am a bit confused by changes like
https://github.com/ovn-org/ovn-kubernetes/commit/c48d85433ed662d261679a12febea741812d0d82#diff-bb96390aa4d292c5ff6e2bc554446182b8abd55e12ba2a0290853b92aa4e1155L429
and then
https://github.com/ovn-org/ovn-kubernetes/commit/92b8cd3f7add1cb845cebfdf923303e98dc91586#diff-bb96390aa4d292c5ff6e2bc554446182b8abd55e12ba2a0290853b92aa4e1155R429
you basically did the change in the first commit and reverted it in the next. Can you please rebase commits so that they have final changes?
Also the commits as they are right now are bit difficult to review - maybe having commit that replaces net.Parse functions with parseSloppy and nothing else, and then next commits for changes related to objects you listed in the description will be easier to review?

I will make these changes and push separate commits.


> Also, can you add some comments for unit test changes you did? I don't see why this [4e83aae](https://github.com/ovn-org/ovn-kubernetes/commit/4e83aaedf2b49815ca43bc05388ad2df3252ec54) change is required - it looks for me like ClusterIP is the right service type

The `Status` field of the `Service` has Loadbalancer Ingress IP set. I updated the `buildServiceLBConfigs` to use `GetExternalAndLBIPs` which adds an additional check on the type of `Service` and only fetches the LB IP if the type is set to `LoadBalancer`. As it was set to `ClusterIP` the unit test was failing.

arkadeepsen commented:
> I will make these changes and push separate commits.

@npinaeva I have pushed 2 separate commits. One for replacing the `net.ParseIP` and `net.ParseCIDR`, and the other for the changes related to the objects.

npinaeva commented:
/lgtm
great work, thanks!

danwinship commented on go-controller/cmd/ovnkube-trace/ovnkube-trace.go:
```diff
@@ -104,14 +105,14 @@ func (pi *PodInfo) String() string {
 
 func (si SvcInfo) getL3Ver() string {
-	if net.ParseIP(si.ClusterIP).To4() != nil {
+	if utilnet.ParseIPSloppy(si.ClusterIP).To4() != nil {
```
You need to use the sloppy functions when parsing k8s API objects, but for other objects, it would be better to make sure that they only ever contain valid IP addresses. So, eg, here, it would be better to change `getSvcInfo` to set `si.ClusterIP` using `ValidateAndGetIPString()`. Or alternatively, make `si.ClusterIP` a `net.IP` rather than a `string`, and set it with `utilnet.ParseIPSloppy(svc.Spec.ClusterIP)`.

(This comment applies everywhere; I'm not going to comment in each place.)

arkadeepsen replied:
Sure. I'll make the appropriate changes. For parsing IPs and CIDRs from the annotation of various objects like namespace, pod, etc., should we use the `sloppy` version of functions or the parse functions from the `net` package?

danwinship replied:
If you have to deal with possibly-bad values in those annotations written by old versions of ovn-kubernetes, then you need to use the sloppy functions. If not, then you don't.

arkadeepsen replied:
I checked the value assignment to the annotations of all the objects and didn't find any bad values (IP or CIDR) being assigned.

Wherever IP strings are being used directly from the k8s API objects I have used `ValidateAndGetIPString` to get the valid IP address first, and then assigned it to other variables or struct fields.

danwinship commented on go-controller/pkg/node/gateway_init_linux_test.go:
```diff
@@ -1202,16 +1203,16 @@ var _ = Describe("Gateway unit tests", func() {
-			_, dpuSubnet, _ := net.ParseCIDR("10.0.0.101/24")
-			nodeIP := net.ParseIP("10.0.1.11")
+			_, dpuSubnet, _ := utilnet.ParseCIDRSloppy("10.0.0.101/24")
+			nodeIP := utilnet.ParseIPSloppy("10.0.1.11")
```
You don't need to use the sloppy functions for const strings that you know are valid.

`net.ParseIP` and `net.ParseCIDR` should still be the default for parsing IPs/CIDRs. The sloppy versions should _only_ be used in contexts where you need to accept invalid strings as user input for backward-compatibility reasons. (Basically, the parameter passed to `ParseIPSloppy` or `ParseCIDRSloppy` should almost always be exactly one of the items in your list in the initial PR description (`service.Spec.ClusterIP`, `endpointSlice.Endpoints[*].Addresses`, etc)

danwinship commented on go-controller/pkg/util/util.go:
```diff
@@ -313,3 +313,15 @@ func UpdateNodeSwitchExcludeIPs(nbClient libovsdbclient.Client, nodeName string,
+
+// ValidateAndGetIPString validates the IP address and returns the IP in
+// string format removing any leading zeros.
+func ValidateAndGetIPString(ip string) (string, error) {
```
If you are only using this function for IPs in kubernetes API objects, then those fields have already been validated, so you know they contain IPs that are valid according to `net.ParseIP`, and it's just a matter of removing the leading zeros. In which case, you don't need to return an `error`.

danwinship reviewed (changes requested):
1. The existing commits should be squashed
2. This appears to have a bunch of "go 1.19 gofmt changes" mixed in with the actual IP parsing changes. It would be better to undo those. If you can't, then this should wait until #3207 merges and then be rebased on top of that.
3. The PR would be a lot simpler if you removed the unused error return from `ValidateAndGetIPString`. If someone needed a version that also works on unvalidated strings later, then they could add that later. You could also rename the function to make it clearer that it requires an already-validated IP string so people don't mistakenly use it on unvalidated IP strings later. But anyway, I'm not going to say you _have_ to change that if you think this way is better.

lgtm with the first two points (at least) fixed, but I feel like someone who is still active in ovn-k development should do the actual merging

arkadeepsen commented:
> 1. The existing commits should be squashed

I have squashed the existing commits.

> 2. This appears to have a bunch of "go 1.19 gofmt changes" mixed in with the actual IP parsing changes. It would be better to undo those. If you can't, then this should wait until #3207 merges and then be rebased on top of that.

I have undone those changes. I also have rebased after #3207 is merged.

> 3. The PR would be a lot simpler if you removed the unused error return from `ValidateAndGetIPString`. If someone needed a version that also works on unvalidated strings later, then they could add that later. You could also rename the function to make it clearer that it requires an already-validated IP string so people don't mistakenly use it on unvalidated IP strings later. But anyway, I'm not going to say you _have_ to change that if you think this way is better.

I have removed the error checking. However, after removing the error checking it just becomes a single line function. So I have removed the function altogether and replaced wherever it was called with `ParseIPSloppy(ip).String()`.
